File: packages/lynx/ios/Sources/HotUpdaterLynxSparklingDiagnostics/SparklingReloadContract.swift
import Foundation

/** Admission and result propagation for a truthful managed-generation reload. */
public enum SparklingReloadContract {
    public static func run(
        closed: Bool,
        replacing: Bool,
        current: Bool,
        error: (_ code: String, _ message: String) -> Error,
        replacement: () -> Result<Void, Error>
    ) -> Result<Void, Error> {
        if closed {
            return .failure(error(
                "HOST_CLOSED",
                "The managed Lynx host is closed"
            ))
        }
        if replacing {
            return .failure(error(
                "RELOAD_BUSY",
                "A managed Lynx generation replacement is already running"
            ))
        }
        if !current {
            return .failure(error(
                "CONTEXT_REJECTED",
                "The requesting Lynx generation is no longer current"
            ))
        }
        return replacement()
    }
}

File: packages/lynx/android-sparkling/src/main/java/com/hotupdater/lynx/sparkling/SparklingReloadContract.kt
package com.hotupdater.lynx.sparkling

import com.hotupdater.lynx.LynxNativeOperationException

/** Admission and result propagation for a truthful managed-generation reload. */
internal object SparklingReloadContract {
    fun run(
        closed: Boolean,
        replacing: Boolean,
        current: Boolean,
        replacement: () -> Result<Unit>,
    ): Result<Unit> {
        val rejection = when {
            closed -> LynxNativeOperationException(
                "HOST_CLOSED",
                "The managed Lynx host is closed",
            )
            replacing -> LynxNativeOperationException(
                "RELOAD_BUSY",
                "A managed Lynx generation replacement is already running",
            )
            !current -> LynxNativeOperationException(
                "CONTEXT_REJECTED",
                "The requesting Lynx generation is no longer current",
            )
            else -> null
        }
        return rejection?.let(Result.Companion::failure) ?: replacement()
    }
}
